mmrazor/models/losses/decoupled_kd.py

```diff
@@ -0,0 +1,140 @@
+        loss_weight: float = 1.0,
+    ) -> None:
+        super(DKDLoss, self).__init__()
+        self.tau = tau
+        accept_reduction = {'none', 'batchmean', 'sum', 'mean'}
+        assert reduction in accept_reduction, \
+            f'KLDivergence supports reduction {accept_reduction}, ' \
+            f'but gets {reduction}.'
+        self.reduction = reduction
+        self.alpha = alpha
+        self.beta = beta
+        self.loss_weight = loss_weight
+
+    def forward(
+        self,
+        student: torch.Tensor,
+        teacher: torch.Tensor,
+        data_samples: list,
+    ) -> torch.Tensor:
+        """DKDLoss forward function.
+
+        Args:
+            student (torch.Tensor): Student featuremap.
+            teacher (torch.Tensor): Teacher featuremap.
+            data_samples (list): List of model input sample data,
+            contains gt_label (dict of label and one-hot score)
+        """
+        # Unpack data samples and pack targets
+        if 'score' in data_samples[0].gt_label:
+            # Batch augmentation may convert labels to one-hot format scores.
+            gt_labels = torch.stack([i.gt_label.score for i in data_samples])
+        else:
+            gt_labels = torch.hstack([i.gt_label.label for i in data_samples])
+        gt_mask = self._get_gt_mask(student, gt_labels)
+        tckd_loss = self._get_tckd_loss(student, teacher, gt_labels, gt_mask)
+        nckd_loss = self._get_nckd_loss(student, teacher, gt_mask)
+        loss = self.alpha * tckd_loss + self.beta * nckd_loss
+        return self.loss_weight * loss
+
+    def _get_nckd_loss(
+        self,
+        student: torch.Tensor,
+        teacher: torch.Tensor,
+        gt_mask: torch.Tensor,
+    ) -> torch.Tensor:
+        """Clac non-target class knowledge distillation."""
+        s_nckd = F.log_softmax(student / self.tau - 1000.0 * gt_mask, dim=1)
+        t_nckd = F.softmax(teacher / self.tau - 1000.0 * gt_mask, dim=1)
+        return self._kl_loss(s_nckd, t_nckd)
+
+    def _get_tckd_loss(
+        self,
+        student: torch.Tensor,
+        teacher: torch.Tensor,
+        gt_labels: torch.Tensor,
+        gt_mask: torch.Tensor,
+    ) -> torch.Tensor:
+        """Calc target class knowledge distillation."""
+        other_mask = self._get_non_gt_mask(student, gt_labels)
+        s_tckd = F.softmax(student / self.tau, dim=1)
+        t_tckd = F.softmax(teacher / self.tau, dim=1)
+        mask_student = torch.log(self._cat_mask(s_tckd, gt_mask, other_mask))
+        mask_teacher = self._cat_mask(t_tckd, gt_mask, other_mask)
+        return self._kl_loss(mask_student, mask_teacher)
+
+    def _kl_loss(
+        self,
+        student: torch.Tensor,
+        teacher: torch.Tensor,
+    ) -> torch.Tensor:
+        kl_loss = F.kl_div(
+            student, teacher, size_average=False,
+            reduction=self.reduction) * self.tau**2
+        return kl_loss
+
+    def _cat_mask(
+        self,
+        target: torch.Tensor,
```

Comment:
It seems that `prob` is better than `target` as it could be the teacher's output probability or the student's one.

Reply:
`Target` stands for the target class knowledge distillation (tckd) of teacher and student, not probe & target.